### PR TITLE
Add DatetimeRangeSlider

### PR DESCRIPTION
**Pull request by ianthomas23**
This is an early-stage WIP to add a `DatetimeRangeSlider`.

Example code using a 1 hour step:
```python
from bokeh.layouts import column
from bokeh.models import CustomJS, DatetimeRangeSlider, Div
from bokeh.plotting import show
from datetime import datetime

slider = DatetimeRangeSlider(
    value=(datetime(2022, 3, 1, 12), datetime(2022, 3, 4, 18)),
    start=datetime(2022, 3, 1, 12),
    end=datetime(2022, 3, 4, 18),
    format = "%d %b %Y %H:%M",
)
div = Div(text="...")

slider.js_on_change(
    "value",
    CustomJS(args=dict(div=div), code="""
div.text=
    'From date: ' + this._formatter(this.value[0], '%d %b %Y') +
    '<br/>From time: ' + this._formatter(this.value[0], '%H:%M') +
    '<br/>To date: ' + this._formatter(this.value[1], '%d %b %Y') +
    '<br/>To time: ' + this._formatter(this.value[1], '%H:%M');
"""))

show(column(slider, div))
```

https://user-images.githubusercontent.com/580326/156781431-558b1eca-914b-4f3b-9de5-c9c194717cb9.mp4

It is similar to `DateRangeSlider`. Has the same API except that it accepts `datetime` objects rather than `date`:
1. `start` and `end` datetimes.
2. Tuple of two `value` datetimes.
3. `format` string.
4. `step` which is in units of the underlying POSIX milliseconds since epoch UTC.

I am showing this early in case anyone has strong views about how it should proceed. Possible improvements:

1. Validation of `step` to restrict it so that an integer multiple of it fits in the next size up date/time unit, e.g. you can't have 27 minutes as your step, but 15, 20, and 30 are OK. Or round it to the nearest acceptable value.
2. Snapping `start`, `end` and the two `value`s to an integer multiple of the `step`.
3. Accepting human-readable strings for the `step`, e.g. "1 hour"?
4. Could add some logic to guess a sensible `step` given the `start` and `end` limits?

### CODE REVIEW

**Comment by jbednar** [comment]
That video looks fabulous! Just what I was hoping to see. We could have used this on so many projects over the years.

> Validation of step to restrict it so that an integer multiple of it fits in the next size up date/time unit, e.g. you can't have 27 minutes as your step, but 15, 20, and 30 are OK. Or round it to the nearest acceptable value.

I'd be permissive about the step, rounding it to conform with whatever requirements Bokeh or the widget itself needs to enforce, but otherwise accepting whatever the user specifies. If people want nice values, they should choose them, but I don't think we should force the values to be nice.

> Snapping start, end and the two values to an integer multiple of the step.

That might be useful in some cases, but I don't think it's desirable in general. Usually the step is just a hint for the GUI, while the start and end can be absolute requirements. E.g. if a dataset covers a certain range of datetimes, it might well be an error to select any range outside those bounds. So I definitely would not want to have the range cover anything other than the precise range the user requested, unless the user explicitly enables such behavior.

> Accepting human-readable strings for the step, e.g. "1 hour"?

Whatever the typical Bokeh behavior is in this respect seems appropriate, e.g. for tick spacing.  E.g. if you can ask for axis ticks spaced "1 hour" apart, then the same here; otherwise doesn't seem appropriate.

> Could add some logic to guess a sensible step given the start and end limits?

I do think there should be a reasonable default step (e.g. `(end-start)/200`); users don't necessarily want to think about the step unless the default isn't appropriate.

File: bokehjs/src/lib/models/widgets/datetime_range_slider.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+  constructor(attrs?: Partial<DatetimeRangeSlider.Attrs>) {
+    super(attrs)
+  }
+
+  static {
+    this.prototype.default_view = DatetimeRangeSliderView
+
+    this.override<DatetimeRangeSlider.Props>({
+      format: "%d %b %Y %H:%M:%S",
+      step: 3600000,  // 1 hour.
```
Comment (mattpap) [inline bokehjs/src/lib/models/widgets/datetime_range_slider.ts:33]: btw. One can use `_` for readability, e.g. `step: 3_600_000,`.

Comment (ianthomas23) [reply bokehjs/src/lib/models/widgets/datetime_range_slider.ts:33]: Thanks.

**Comment by ianthomas23** [comment]
Based on @jbednar's comments, we can take the view that the user needs to specify reasonable values for all of `start`, `end`, `value` and `step` and we don't need to do anything clever guessing reasonable values that are not specified. So I will continue with adding tests and examples for this.

**Comment by jbednar** [comment]
I think step does warrant a default, because users won't necessarily have any opinion about its value until they try it. But I think its default should be whatever other Bokeh sliders default to, presumably some fixed divison of the range.

**Comment by ianthomas23** [comment]
This is ready for review. The Windows test failure doesn't appear to be related?

**Comment by philippjfr** [comment]
This looks great to me and will be very useful!

**Comment by bryevdv** [comment]
OMG after however many years, GitHub has finally, finally added "Re-run *failed jobs*"

**Review by bryevdv — APPROVED** [review]

**Pull request merged**